### PR TITLE
Enumerable

### PR DESCRIPTION
**Pull request by Hermela-Fikre**
In this project, i have done the following:

    Added my new enumerable methods into the existing enumerable module
    Tested all my methods to ensure they work as expected
    Used rubocop to enforce development guidelines

### CODE REVIEW

File: .rubocop.yml [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+Lint/StructNewOverride:
+  Enabled: true
+Style/HashEachMethods:
+  Enabled: true 
+
+
+Style/HashTransformKeys:
+  Enabled: true
+Style/HashTransformValues:
+  Enabled: true
```
Comment (shshamim63) [inline .rubocop.yml:32]: You should not alter the provided config file. But if you need to do it for any reason you should inform it in the PR description. Check-out this [link](https://github.com/microverseinc/linters-config/tree/master/ruby). 
basic instruction

File: main.rb [hunk 2]
```diff
@@ -0,0 +1,93 @@
+module Enumerable
+  def my_each
+    size.times do |i|
+      yield self[i]
+    end
+  end
```
Comment (shshamim63) [inline main.rb]: `my_each` identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) ``
- Should return an enumerator if no block given.

File: main.rb [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_each_with_index
+    size.times { |i| yield self[i], i }
+  end
```
Comment (shshamim63) [inline main.rb]: `my_each_with_index` identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- Should return an enumerator if no block given.

File: main.rb [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_select
+    result_array = []
+    my_each do |i|
+      result_array << i if yield(i)
+    end
+    result_array
+  end
```
Comment (shshamim63) [inline main.rb]: `my_select` identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- Should return an enumerator if no block given.

File: main.rb [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_all?
+    result = true
+    my_each { |i| result = false unless yield(i) }
+    result
+  end
```
Comment (shshamim63) [inline main.rb]: `my_all?` identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2)) 
- when no block or argument is given, returns true when none of the collection members are false or nil
```ruby
[1, 2, 3].my_all? #should return true
[1, 2, nil].my_all? #should return false
[1, false, nil].my_all? #should return false
```
- when a class is passed as an argument returns true if all of the collection is a member of such class
```ruby
[1, 2, 3].my_all?(Integer) #should return true
[1, 2, nil].my_all?(Integer) #should return false
```
- when a Regex is passed as an argument returns true if all of the collection matches the Regex
```ruby
%w[dog door rod blade].my_all?(/d/)   #should return true.
%w[dog door rod blade].my_all?(/t/)   #should return false.
```
- when a pattern other than Regex or a Class is given, returns true if all of the collection matches the pattern
```ruby
[3, 3, 3].my_all?(3) #should return true
[3, 3, 1].my_all?(3) #should return false
```

File: main.rb [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_any?
+    result = false
+    my_each { |i| result = true if yield(i) }
+    result
+  end
```
Comment (shshamim63) [inline main.rb]: `my_any?` identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when no block or argument is given returns true if at least one of the collection is not false or nil
```ruby
[1, 2, 3].my_any? #should return true
[1, 2, nil].my_any? #should return true
[false, false, nil].my_any? #should return false
```
- when a class is passed as an argument returns true if at least one of the collection is a member of such class
```ruby
[1, 2, 3].my_any?(Integer) #should return true
[1, 'demo', false].my_ant?(Integer) #should return true
['demo', false, nil].my_any?(Integer) #should return false
```
- when a Regex is passed as an argument returns false if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_any?(/t/)    #should return false.
%w[dog door rod blade].my_any?(/d/)   #should return true.
```
- when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
```ruby
[1, 2, 3].my_any?(3)   #should return true
[1, 2, 2].my_any?(3)   #should return false
```

File: main.rb [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_none?
+    result = true
+    my_each { |i| result = false if yield(i) }
+    result
+  end
```
Comment (shshamim63) [inline main.rb]: `my_none?` identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- when no block or argument is given returns true only if none of the collection members is true
```ruby
[false, nil, false].my_none? #should return true
[1, 'demo', 2.2].my_none? #should return false
```
- when a class is passed as an argument returns true if none of the collection is a member of such class
```ruby
[1, 2, 3].my_none?(String)  #should return true
['1', '2', '3'].my_none?(String)  #should return false
```
- when a Regex is passed as an argument returns true only if none of the collection matches the Regex
```ruby
%w[dog door rod blade].my_none?(/z/)   #should return true
%w[dog door rod blade].my_none?(/d/)   #should return false
```
- when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
```ruby
[1, 2, 4].my_none?(3) #should return true
[1, 2, 3].my_none?(3) #should return false
```

File: main.rb [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_count
+    result = 0
+    my_each { |i| result += 1 if yield(i) }
+    result
+  end
```
Comment (shshamim63) [inline main.rb]: `my_count` identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns the number of items in enum through enumeration
```ruby
[1, 2, 3].my_count  #should return 3
```
- counts the number of items in enum that are equal to item if an argument is given
```ruby
[1, 1, 2 ,3].my_count(1)  #should return 2
```

File: main.rb [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+  def my_map(num = false)
+    result = []
+    if num == false
+      my_each { |i| result.push(yield(i)) }
+      result
+    else
+      my_each { |i| result.push num.call(i) }
+    end
+    result
+  end
```
Comment (shshamim63) [inline main.rb]: `my_map` identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- returns an Enumerator if no block is given

File: main.rb [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+  def my_inject
+    result = first
+    my_each_with_index do |item, index|
+      next if index.zero?
+
+      result = yield(result, item)
+    end
+    result
+  end
```
Comment (shshamim63) [inline main.rb]: `my_inject` identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- combines all elements of enum by applying a binary operation, specified by a block
```ruby
[1, 2, 3].my_inject { |memo, num| memo + num } #should return 6
[1, 2, 3].my_inject(1) { |memo, num| memo + num } #should return 7
```
- when a symbol is specified combines each element of the collection by applying the symbol as a named method
```ruby
[1, 2, 3].my_inject(:+)  #should return 6
[1, 2, 3].my_inject(1, :+)  #should return 7
```

File: main.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+
+  def multiply_els
+    my_inject { |total, item| total * item }
+  end
```
Comment (shshamim63) [inline main.rb]: multiply_els is not an enumerable method. So it should not be inside the module. And the functionality of it specifies that,
- should accept an array as a parameter
- performs inject operation on the array.
So it should be something like,
```ruby
def multiply_els(arr)
  arr.my_inject(1) { |prod, num| prod * num }
end
```

File: main.rb [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+
+# puts([1, 2, 3, 4, 5].my_map { |item| item * 2 })
+
+# my_proc = proc { |item| item * 5 }
+
+# puts [1, 2, 3, 4, 5].my_map(my_proc)
+
+# puts [2, 2, 3, 4, 5].my_inject { |total, element| total * element }
+
+# puts [2, 4, 5].multiply_els
```
Comment (shshamim63) [inline main.rb]: This is not the way we do test. You will implement the real test in one of your upcoming projects. To make your code clean there should not be any commented code. So I will believe it is better to delete them. If you want to check a method well call it with puts but don't commit it the puts related code.

**Review by shshamim63 — CHANGES_REQUESTED** [review]
## Status :traffic_light: Changes Required :tractor: 
___
Hi @Hermela-Fikre :wave: ,
Great initiative. The code looks neat and clean but needs some refactoring that you can do as you proceed with it. I have found some issues that are holding me to approve it in this stage. But the amazing thing is, based on your coding skill I believe it won't be a problem for you to complete. So let's hop on to it.
___
### :spades: Mandatory Changes :spades: 
___
Check out the comments below, I have found issues with them. So you fix a method try to execute them with the given samples. It will help you to find the actual problem.
___
##### Please fix the issues and send another review request.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

**Comment by Anna-Myzukina** [comment]
## Hi @Hermela-Fikre  !  :wave: 

# You did a great job, approved  :+1:

### Go and merge to master.
`Goodluck in your next milestones Happy coding ` <g-emoji class="g-emoji" alias="computer" fallback-src="https://github.githubassets.com/images/icons/emoji/unicode/1f4bb.png">computer</g-emoji>

**Pull request merged**